_posts/2025-04-24-secure-mcp-sse-server.adoc

```diff
@@ -0,0 +1,227 @@
+
+No problems if you do not work with Keycloak, see the <<mcp-server-devui>> section for more details.
+
+We are ready to test our secure MCP server.
+
+== Start MCP SSE server
+
+Start the server in the dev mode:
+
+[source,shell]
+----
+mvn quarkus:dev
+----
+
+== Access MCP SSE server
+
+[[mcp-server-devui]]
+=== Use Quarkus MCP Server Dev UI
+
+Go to http://localhost:8080/q/dev[Dev UI], find both MCP Server and OpenId Connect cards:
+
+image::mcp_server_oidc_devui.png[MCP Server and OIDC in DevUI,align="center"]
+
+Select an OpenId Connect card and https://quarkus.io/guides/security-openid-connect-dev-services#develop-service-applications[login to Keycloak] using an `alice` name and an `alice` password.
+
+[NOTE]
+====
+You can login to other providers such as `Auth0` or https://quarkus.io/guides/security-openid-connect-providers#github[GitHub] from OIDC DevUI as well. The only requirement is to update your application registration to allow callbacks to DevUI. For example, see how you can https://quarkus.io/guides/security-oidc-auth0-tutorial#looking-at-auth0-tokens-in-the-oidc-dev-ui[login to Auth0 from Dev UI].
+
+After logging in with `Keycloak` as `alice`, copy the acquired access token using a provided copy button:
+
+image::login_and_copy_access_token.png[Login and copy access token,align="center"]
+
+Next, go to the MCP Server card, select the `Tools` option and choose to `Call` the `user-name-provider` tool:
+
+image::mcp_server_choose_tool.png[Choose MCP Server tool,align="center"]
+
+Paste the copied access token into the Tool's `Bearer token` field, and request a new MCP SSE session:
+
+image::mcp_server_bearer_token.png[MCP Server Bearer token,align="center"]
+
+Use the `Call` action to make a tool call and get a response which contains the `alice` user name:
+
+image::mcp_server_tool_response.png[MCP Server tool response,align="center"]
+
+[[mcp-inspector]]
+=== Use MCP Inspector
+
+Launch https://modelcontextprotocol.io/docs/tools/inspector[MCP inspector]:
+
+[source,shell]
+----
+npx @modelcontextprotocol/inspector
+----
+
+Next, get and copy the access token as explained in the <<mcp-server-devui>> section above. You can use the same access token that you may have already used when working through that section.
+
+Use this access token to have MCP inspector connect to the Quarkus MCP SSE server:
+
+image::mcp_inspector_connect.png[MCP Inspector Connect,align="center"]
+
+Next, make a `user-name-provider` tool call:
+
+image::mcp_inspector_tool_call.png[MCP Inspector Tool Call,align="center"]
+
+=== Use curl
+
+Finally, let's use `curl` and also learn a little bit how the MCP protocol and MCP SSE transport work. 
+
```

Comment:
What about adding something like this (and a picture):

When invoking an MCP tool using the SSE transport, there are at least two requests:

1. A first request connects to the server using SSE - this _back channel_ receives the invocation results and notifications
2. A second request (`POST`) invokes the tool. The invocation result is transmitted on the SSE _back channel_.

Reply:
@cescoffier Sorry, missed it too, good idea, if you still think it is worth adding it, please create a diagram with your nice tools, may be on Monday ? and We  can add it to the `curl` section, I can try to sketch it as well, but you have a very clear picture how it should be presented, so yeah, please do it and we can add it tomorrow